File: project-bolt-sb1-7qmsjyym/project/src/constraints/ConstraintValidator.ts
import { Teacher, ClassGroup, Schedule, Assignment, TimeSlot } from '../models';

export interface ValidationResult {
  isValid: boolean;
  errors: string[];
  warnings: string[];
}

export class ConstraintValidator {
  
  /**
   * בדיקת תקינות כללית של המערכת
   */
  validateSystem(
    teachers: Teacher[], 
    classes: ClassGroup[], 
    schedule: Schedule
  ): ValidationResult {
    const errors: string[] = [];
    const warnings: string[] = [];

    // בדיקת תקינות מורות
    teachers.forEach((teacher, index) => {
      const teacherErrors = this.validateTeacher(teacher);
      errors.push(...teacherErrors.map(e => `Teacher ${index + 1}: ${e}`));
    });

    // בדיקת תקינות כיתות
    classes.forEach((classGroup, index) => {
      const classErrors = this.validateClass(classGroup);
      errors.push(...classErrors.map(e => `Class ${index + 1}: ${e}`));
    });

    // בדיקת תקינות מערכת שעות
    const scheduleErrors = this.validateSchedule(schedule, teachers, classes);
    errors.push(...scheduleErrors);

    // בדיקות אזהרה
    const systemWarnings = this.generateWarnings(teachers, classes, schedule);
    warnings.push(...systemWarnings);

    return {
      isValid: errors.length === 0,
      errors,
      warnings
    };
  }

  /**
   * בדיקת תקינות מורה בודדת
   */
  private validateTeacher(teacher: Teacher): string[] {
    const errors: string[] = [];
    
    if (!teacher.id?.trim()) {
      errors.push('Teacher ID is required');
    }
    
    if (!teacher.name?.trim()) {
      errors.push('Teacher name is required');
    }
    
    if (!teacher.subjects || teacher.subjects.length === 0) {
      errors.push('Teacher must have at least one subject');
    }
    
    teacher.subjects?.forEach((subject, index) => {
      if (!subject.name?.trim()) {
        errors.push(`Subject ${index + 1}: name is required`);
      }
      if (subject.gradeLevel < 1 || subject.gradeLevel > 8) {
        errors.push(`Subject ${index + 1}: grade level must be between 1-8`);
      }
    });
    
    if (teacher.preferredDayOff !== undefined && 
        (teacher.preferredDayOff < 0 || teacher.preferredDayOff > 6)) {
      errors.push('Preferred day off must be between 0-6');
    }
    
    return errors;
  }

  /**
   * בדיקת תקינות כיתה בודדת
   */
  private validateClass(classGroup: ClassGroup): string[] {
    const errors: string[] = [];
    
    if (!classGroup.id?.trim()) {
      errors.push('Class ID is required');
    }
    
    if (!classGroup.name?.trim()) {
      errors.push('Class name is required');
    }
    
    if (classGroup.gradeLevel < 1 || classGroup.gradeLevel > 8) {
      errors.push('Grade level must be between 1-8');
    }
    
    if (classGroup.minHoursPerDay < 0) {
      errors.push('Minimum hours per day cannot be negative');
    }
    
    if (classGroup.maxHoursPerDay < classGroup.minHoursPerDay) {
      errors.push('Maximum hours per day must be >= minimum hours per day');
    }
    
    if (!classGroup.subjectRequirements || classGroup.subjectRequirements.length === 0) {
      errors.push('Class must have at least one subject requirement');
    }
    
    classGroup.subjectRequirements?.forEach((req, index) => {
      if (!req.subject?.trim()) {
        errors.push(`Subject requirement ${index + 1}: subject name is required`);
      }
      if (req.hoursPerWeek <= 0) {
        errors.push(`Subject requirement ${index + 1}: hours per week must be positive`);
      }
    });
    
    return errors;
  }

  /**
   * בדיקת תקינות מערכת שעות
   */
  private validateSchedule(
    schedule: Schedule, 
    teachers: Teacher[], 
    classes: ClassGroup[]
  ): string[] {
    const errors: string[] = [];
    const assignments = schedule.getAllAssignments();

    // בדיקת התנגשויות
    for (let i = 0; i < assignments.length; i++) {
      for (let j = i + 1; j < assignments.length; j++) {
        if (this.assignmentsConflict(assignments[i], assignments[j])) {
          errors.push(
            `Conflict: ${assignments[i].teacherId} and ${assignments[j].teacherId} ` +
            `at day ${assignments[i].timeSlot.day}, hour ${assignments[i].timeSlot.hour}`
          );
        }
      }
    }

    // בדיקת התאמת מורה למקצוע ושכבה
    assignments.forEach((assignment, index) => {
      const teacher = teachers.find(t => t.id === assignment.teacherId);
      const classGroup = classes.find(c => c.id === assignment.classId);
      
      if (!teacher) {
        errors.push(`Assignment ${index + 1}: Teacher ${assignment.teacherId} not found`);
        return;
      }
      
      if (!classGroup) {
        errors.push(`Assignment ${index + 1}: Class ${assignment.classId} not found`);
        return;
      }
      
      const canTeach = teacher.subjects.some(s => 
        s.name === assignment.subject && s.gradeLevel === classGroup.gradeLevel
      );
      
      if (!canTeach) {
        errors.push(
          `Assignment ${index + 1}: Teacher ${teacher.name} cannot teach ` +
          `${assignment.subject} to grade ${classGroup.gradeLevel}`
        );
      }
    });

    return errors;
  }

  /**
   * בדיקת התנגשות בין שני שיבוצים
   */
  private assignmentsConflict(a1: Assignment, a2: Assignment): boolean {
    const sameTime = a1.timeSlot.day === a2.timeSlot.day && 
                     a1.timeSlot.hour === a2.timeSlot.hour;
    
    if (!sameTime) return false;
    
    // מורה לא יכולה להיות בשני מקומות
    if (a1.teacherId === a2.teacherId) return true;
    
    // כיתה לא יכולה להיות עם שני מורים
    if (a1.classId === a2.classId) return true;
    
    return false;
  }

  /**
   * יצירת אזהרות למערכת
   */
  private generateWarnings(
    teachers: Teacher[], 
    classes: ClassGroup[], 
    schedule: Schedule
  ): string[] {
    const warnings: string[] = [];

    // אזהרה על מורות עם עומס גבוה
    teachers.forEach(teacher => {
      const assignments = schedule.getAssignmentsForTeacher(teacher.id);
      if (assignments.length > 30) { // יותר מ-30 שעות בשבוע
        warnings.push(`Teacher ${teacher.name} has ${assignments.length} hours per week (high load)`);
      }
    });

    // בדיקת השלמת דרישות כיתות - אילוץ קשיח
    classes.forEach(classGroup => {
      // בדיקה לכל מקצוע בנפרד
      classGroup.subjectRequirements.forEach(requirement => {
        const assignments = schedule.getAssignmentsForClass(classGroup.id)
          .filter(a => a.subject === requirement.subject);
        
        if (assignments.length !== requirement.hoursPerWeek) {
          warnings.push(
            `Class ${classGroup.name}: ${requirement.subject} has ${assignments.length}/${requirement.hoursPerWeek} hours`
          );
        }
      });
      
      // בדיקת מגבלות יומיות
      const workingDays = [0, 1, 2, 3, 4];
      workingDays.forEach(day => {
        const dailyHours = schedule.getClassHoursPerDay(classGroup.id, day);
        
        // אזהרה על חריגה ממקסימום
        if (dailyHours > classGroup.maxHoursPerDay) {
          warnings.push(
            `Class ${classGroup.name} exceeds max hours on day ${day}: ${dailyHours}/${classGroup.maxHoursPerDay}`
          );
        }
        
        // אזהרה על פחות ממינימום
        if (dailyHours > 0 && dailyHours < classGroup.minHoursPerDay) {
          warnings.push(
            `Class ${classGroup.name} below min hours on day ${day}: ${dailyHours}/${classGroup.minHoursPerDay}`
          );
        }
      });
    });

    return warnings;
  }
}

File: project-bolt-sb1-7qmsjyym/project/python_solver/constraints/constraint_validator.py
from typing import List
from dataclasses import dataclass

from models.teacher import Teacher
from models.class_group import ClassGroup
from models.schedule import Schedule, Assignment

@dataclass
class ValidationResult:
    is_valid: bool
    errors: List[str]
    warnings: List[str]

class ConstraintValidator:
    
    def validate_system(self, teachers: List[Teacher], classes: List[ClassGroup], 
                       schedule: Schedule) -> ValidationResult:
        """בדיקת תקינות כללית של המערכת"""
        errors = []
        warnings = []
        
        # בדיקת תקינות מורות
        for i, teacher in enumerate(teachers):
            teacher_errors = self._validate_teacher(teacher)
            errors.extend([f'Teacher {i + 1}: {e}' for e in teacher_errors])
        
        # בדיקת תקינות כיתות
        for i, class_group in enumerate(classes):
            class_errors = self._validate_class(class_group)
            errors.extend([f'Class {i + 1}: {e}' for e in class_errors])
        
        # בדיקת תקינות מערכת שעות
        schedule_errors = self._validate_schedule(schedule, teachers, classes)
        errors.extend(schedule_errors)
        
        # בדיקות אזהרה
        system_warnings = self._generate_warnings(teachers, classes, schedule)
        warnings.extend(system_warnings)
        
        return ValidationResult(
            is_valid=len(errors) == 0,
            errors=errors,
            warnings=warnings
        )
    
    def _validate_teacher(self, teacher: Teacher) -> List[str]:
        """בדיקת תקינות מורה בודדת"""
        errors = []
        
        if not teacher.id or not teacher.id.strip():
            errors.append('Teacher ID is required')
        
        if not teacher.teacher_name or not teacher.teacher_name.strip():
            errors.append('Teacher name is required')
        
        if not teacher.subjects:
            errors.append('Teacher must have at least one subject')
        
        for i, subject in enumerate(teacher.subjects):
            if not subject.subject_name or not subject.subject_name.strip():
                errors.append(f'Subject {i + 1}: name is required')
            if subject.grade_level < 1 or subject.grade_level > 8:
                errors.append(f'Subject {i + 1}: grade level must be between 1-8')
        
        if (teacher.preferred_day_off is not None and 
            (teacher.preferred_day_off < 0 or teacher.preferred_day_off > 6)):
            errors.append('Preferred day off must be between 0-6')
        
        return errors
    
    def _validate_class(self, class_group: ClassGroup) -> List[str]:
        """בדיקת תקינות כיתה בודדת"""
        errors = []
        
        if not class_group.id or not class_group.id.strip():
            errors.append('Class ID is required')
        
        if not class_group.class_name or not class_group.class_name.strip():
            errors.append('Class name is required')
        
        if class_group.grade_level < 1 or class_group.grade_level > 8:
            errors.append('Grade level must be between 1-8')
        
        if class_group.min_hours_per_day < 0:
            errors.append('Minimum hours per day cannot be negative')
        
        if class_group.max_hours_per_day < class_group.min_hours_per_day:
            errors.append('Maximum hours per day must be >= minimum hours per day')
        
        if not class_group.subject_requirements:
            errors.append('Class must have at least one subject requirement')
        
        for i, req in enumerate(class_group.subject_requirements):
            if not req.subject or not req.subject.strip():
                errors.append(f'Subject requirement {i + 1}: subject name is required')
            if req.hours_per_week <= 0:
                errors.append(f'Subject requirement {i + 1}: hours per week must be positive')
        
        return errors
    
    def _validate_schedule(self, schedule: Schedule, teachers: List[Teacher], 
                          classes: List[ClassGroup]) -> List[str]:
        """בדיקת תקינות מערכת שעות"""
        errors = []
        assignments = schedule.get_all_assignments()
        
        # בדיקת התנגשויות
        for i in range(len(assignments)):
            for j in range(i + 1, len(assignments)):
                if self._assignments_conflict(assignments[i], assignments[j]):
                    errors.append(
                        f'Conflict: {assignments[i].teacher_id} and {assignments[j].teacher_id} '
                        f'at day {assignments[i].time_slot._day}, hour {assignments[i].time_slot._hour}'
                    )
        
        # בדיקת התאמת מורה למקצוע ושכבה
        for i, assignment in enumerate(assignments):
            teacher = next((t for t in teachers if t.id == assignment.teacher_id), None)
            class_group = next((c for c in classes if c.id == assignment.class_id), None)
            
            if not teacher:
                errors.append(f'Assignment {i + 1}: Teacher {assignment.teacher_id} not found')
                continue
            
            if not class_group:
                errors.append(f'Assignment {i + 1}: Class {assignment.class_id} not found')
                continue
            
            can_teach = any(s.subject_name == assignment.subject and s.grade_level == class_group.grade_level
                           for s in teacher.subjects)
            
            if not can_teach:
                errors.append(
                    f'Assignment {i + 1}: Teacher {teacher.teacher_name} cannot teach '
                    f'{assignment.subject} to grade {class_group.grade_level}'
                )
        
        # בדיקת חורים במערכת
        gaps = schedule.get_schedule_gaps()
        for class_id, gap_days in gaps.items():
            for day in gap_days:
                errors.append(f'Class {class_id} has gaps in schedule on day {day}')
        
        return errors
    
    def _assignments_conflict(self, a1: Assignment, a2: Assignment) -> bool:
        """בדיקת התנגשות בין שני שיבוצים"""
        same_time = (a1.time_slot._day == a2.time_slot._day and
                    a1.time_slot._hour == a2.time_slot._hour)
        
        if not same_time:
            return False
        
        # מורה לא יכולה להיות בשני מקומות
        if a1.teacher_id == a2.teacher_id:
            return True
        
        # כיתה לא יכולה להיות עם שני מורים
        if a1.class_id == a2.class_id:
            return True
        
        return False
    
    def _generate_warnings(self, teachers: List[Teacher], classes: List[ClassGroup], 
                          schedule: Schedule) -> List[str]:
        """יצירת אזהרות למערכת"""
        warnings = []
        
        # אזהרה על מורות עם עומס גבוה
        for teacher in teachers:
            assignments = schedule.get_assignments_for_teacher(teacher.id)
            if len(assignments) > 30:  # יותר מ-30 שעות בשבוע
                warnings.append(f'Teacher {teacher.teacher_name} has {len(assignments)} hours per week (high load)')
        
        # בדיקת השלמת דרישות כיתות
        for class_group in classes:
            for requirement in class_group.subject_requirements:
                assignments = [
                    a for a in schedule.get_assignments_for_class(class_group.id)
                    if a.subject == requirement.subject
                ]
                
                if len(assignments) != requirement.hours_per_week:
                    warnings.append(
                        f'Class {class_group.class_name}: {requirement.subject} has '
                        f'{len(assignments)}/{requirement.hours_per_week} hours'
                    )
            
            # בדיקת מגבלות יומיות
            working_days = [0, 1, 2, 3, 4]
            for day in working_days:
                daily_hours = schedule.get_class_hours_per_day(class_group.id, day)
                
                if daily_hours > class_group.max_hours_per_day:
                    warnings.append(
                        f'Class {class_group.class_name} exceeds max hours on day {day}: '
                        f'{daily_hours}/{class_group.max_hours_per_day}'
                    )
                
                if daily_hours > 0 and daily_hours < class_group.min_hours_per_day:
                    warnings.append(
                        f'Class {class_group.class_name} below min hours on day {day}: '
                        f'{daily_hours}/{class_group.min_hours_per_day}'
                    )
        
        return warnings